File: cs-java/src/test/java/control/structures/examples/GivenAnEuclideanPoint.java
package control.structures.examples;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.FieldSource;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import static org.junit.jupiter.api.Assertions.*;
import static org.hamcrest.number.IsCloseTo.*;
import static org.hamcrest.MatcherAssert.*;


public class GivenAnEuclideanPoint {
    @SuppressWarnings("unused")
    private static final Logger logger = LoggerFactory.getLogger(GivenAnEuclideanPoint.class);

    public interface IPythagoras {
        double pythagoras(double x, double y);
    }

    static Pythagoras pythagoras = new Pythagoras();
    static double x = 3.0;
    static double y = 4.0;
    static double expected_result = 5.0;

    static IPythagoras[] whenCalculatingThePythagorasThenTheResultIsTheDistanceFromTheOrigin = {
            (IPythagoras) (double x, double y) -> pythagoras.pythagoras_function_call_with_local_variables(x,y),
            (IPythagoras) (double x, double y) -> pythagoras.pythagoras_function_calls_per_level(x,y),
            (IPythagoras) (double x, double y) -> pythagoras.pythagoras_function_calls_per_level_by_need(x,y),
            (IPythagoras) (double x, double y) -> pythagoras.pythagoras_continuation_passing_style(x,y),
    };

    @ParameterizedTest
    @FieldSource
    public void whenCalculatingThePythagorasThenTheResultIsTheDistanceFromTheOrigin(IPythagoras ipythagoras) {
        // IPythagoras ipythagoras = (IPythagoras) (double x, double y) -> pythagoras.pythagoras_function_call_with_local_variables(x,y);
        assertThat("The distance of the point (3.0, 4.0) from the origin should be 5.0.", ipythagoras.pythagoras(x, y), closeTo(expected_result, 0.01));
    }

}

File: app/src/test/kotlin/control/structures/kt/examples/GivenAnEuclideanPoint.kt
package control.structures.kt.examples

import org.hamcrest.MatcherAssert.*
import org.hamcrest.number.IsCloseTo.*
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.FieldSource
import org.slf4j.Logger
import org.slf4j.LoggerFactory

fun interface IPythagoras {
    fun pythagoras(x: Double, y: Double): Double
}

class GivenAnEuclideanPoint {
    companion object {
        val logger: Logger = LoggerFactory.getLogger(GivenAnEuclideanPoint::class.java.name)

        var pythagoras: Pythagoras = Pythagoras()
        var x: Double = 3.0
        var y: Double = 4.0
        var expected_result: Double = 5.0

        val whenCalculatingThePythagorasThenTheResultIsTheDistanceFromTheOrigin: Array<IPythagoras> = arrayOf(
            IPythagoras {x: Double, y: Double -> pythagoras.pythagoras_communicating_sequential_processes_style(x,y) },
            IPythagoras {x: Double, y: Double -> pythagoras.pythagoras_kotlin_flow(x, y)},
        )
    }


    @ParameterizedTest
    @FieldSource
    fun whenCalculatingThePythagorasThenTheResultIsTheDistanceFromTheOrigin(ipythagoras: IPythagoras) {
        // IPythagoras ipythagoras = (IPythagoras) (double x, double y) -> pythagoras.pythagoras_function_call_with_local_variables(x,y);
        assertThat<Double>("The distance of the point (3.0, 4.0) from the origin should be 5.0.", ipythagoras.pythagoras(x, y),  closeTo(expected_result, 0.01))
    }

}